pgjdbc/src/main/java/org/postgresql/jdbc/PgResultSet.java

```diff
@@ -3271,90 +3272,164 @@ public void updateArray(String columnName, Array x) throws SQLException {
           return null;
         }
         return type.cast(doubleValue);
-      } else {
-        throw new PSQLException(GT.tr("conversion to {0} from {1} not supported", type, sqlType),
-                PSQLState.INVALID_PARAMETER_VALUE);
       }
-    } else if (type == Date.class) {
-      if (sqlType == Types.DATE) {
-        return type.cast(getDate(columnIndex));
-      } else {
-        throw new PSQLException(GT.tr("conversion to {0} from {1} not supported", type, sqlType),
-                PSQLState.INVALID_PARAMETER_VALUE);
-      }
-    } else if (type == Time.class) {
-      if (sqlType == Types.TIME) {
-        return type.cast(getTime(columnIndex));
-      } else {
-        throw new PSQLException(GT.tr("conversion to {0} from {1} not supported", type, sqlType),
-                PSQLState.INVALID_PARAMETER_VALUE);
-      }
-    } else if (type == Timestamp.class) {
+    } else if (DATE_CLASSES.contains(type)) {
       if (sqlType == Types.TIMESTAMP
-              //#if mvn.project.property.postgresql.jdbc.spec >= "JDBC4.2"
-              || sqlType == Types.TIMESTAMP_WITH_TIMEZONE
+          //#if mvn.project.property.postgresql.jdbc.spec >= "JDBC4.2"
+          || sqlType == Types.TIMESTAMP_WITH_TIMEZONE
       //#endif
       ) {
-        return type.cast(getTimestamp(columnIndex));
-      } else {
-        throw new PSQLException(GT.tr("conversion to {0} from {1} not supported", type, sqlType),
-                PSQLState.INVALID_PARAMETER_VALUE);
+        final Timestamp timestampValue = getTimestamp(columnIndex);
+        if (timestampValue == null) {
+          return null;
+        }
+        if (type == Timestamp.class) {
+          return type.cast(timestampValue);
+        }
+
+        final long time = timestampValue.getTime();
+
+        if (type == java.util.Date.class) {
+          return type.cast(new java.util.Date(time));
+        }
+
+        if (type == Calendar.class) {
+          final Calendar calendar = Calendar.getInstance(getDefaultCalendar().getTimeZone());
```

Comment:
to be consistent with ZonedDateTime and OffsetDateTime, should this flex which timezone is used based on whether TIMESTAMP or TIMESTAMPTZ?
There is currently inconsistency between Calendar and OffsetDateTime. Do we prefer this continued inconsistency or a non-passive change to handling of Calendar?